mlflow/langchain/api_request_parallel_processor.py

```diff
@@ -107,23 +108,64 @@ class APIRequest:
-        # update self.request_json to make sure error message is accurate
-        self.request_json, did_perform_chat_conversion = (
-            transform_request_json_for_chat_if_necessary(single_input, self.lc_model)
-        )
-
-        if self.stream:
-            result = self.lc_model.stream(self.request_json, config=config, **kwargs)
+    def _call_model(
+        self,
+        input,
+        config: dict,
+        callback_handlers: Optional[list[BaseCallbackHandler]],
+        stream: bool,
+        **kwargs,
+    ):
+        if stream:
+            return self.lc_model.stream(input, config=config, **kwargs)
         else:
             if hasattr(self.lc_model, "invoke"):
-                result = self.lc_model.invoke(self.request_json, config=config, **kwargs)
+                return self.lc_model.invoke(input, config=config, **kwargs)
             else:
                 # for backwards compatibility, __call__ is deprecated and will be removed in 0.3.0
                 # kwargs shouldn't have config field if invoking with __call__
-                result = self.lc_model(self.request_json, callbacks=callback_handlers, **kwargs)
+                return self.lc_model(input, callbacks=callback_handlers, **kwargs)
+
+    def _predict_single_input(
+        self, callback_handlers: Optional[list[BaseCallbackHandler]], **kwargs
+    ):
+        config = kwargs.pop("config", {})
+        config["callbacks"] = config.get("callbacks", []) + (callback_handlers or [])
+        if json_dict_might_be_chat_request(self.request_json):
+            # Try to convert the request to list[BaseMessage] type and call the model
+            # If the conversion fails, call the model with the original request
+            # Note: we cannot rely on the model's input_schema to validate the request
+            # because lots of cases the input_schema is not reliable.
+            try:
+                request_json = _convert_chat_request_to_messages_or_throw(self.request_json)
+                if self.stream:
+                    # validate if the model accepts list[BaseMessage] as input
+                    # as stream doesn't calculate the result immediately
+                    self._call_model(
+                        request_json, config, callback_handlers, stream=False, **kwargs
+                    )
+                result = self._call_model(
+                    request_json, config, callback_handlers, stream=self.stream, **kwargs
+                )
+                # only update after above execution succeeds to make sure error message is correct
+                self.request_json = request_json
+                did_perform_chat_conversion = True
+            # it is hard to catch specific exceptions here because langchain models
+            # don't raise common input validation errors
+            # TODO: what if the model accepts the messages input but just fails?
```

Comment:
This is the risk though, but I haven't come up with a better solution. In the old ways, I think we just convert the request to list[message] and try to invoke with it directly, which is also not accurate